### PR TITLE
User registration

### LINKED ISSUE #9: User Story 12
User Story 12, Registration Email must be unique

As a visitor
When I visit the user registration page
If I fill out the registration form
But include an email address already in the system
Then I am returned to the registration page
My details are not saved and I am not logged in
The form is filled in with all previous data except the email field and password fields
I see a flash message telling me the email address is already in use

### LINKED ISSUE #10: User Story 11
User Story 11, User Registration Missing Details

As a visitor
When I visit the user registration page
And I do not fill in this form completely,
I am returned to the registration page
And I see a flash message indicating that I am missing required fields

### LINKED ISSUE #11: User Story 10
User Story 10, User Registration

As a visitor
When I click on the 'register' link in the nav bar
Then I am on the user registration page ('/register')
And I see a form where I input the following data:
- my name
- my street address
- my city
- my state
- my zip code
- my email address
- my preferred password
- a confirmation field for my password

When I fill in this form completely,
And with a unique email address not already in the system
My details are saved in the database
Then I am logged in as a registered user
I am taken to my profile page ("/profile")
I see a flash message indicating that I am now registered and logged in

### PR DESCRIPTION
**Pull request by dcoleman21**
Completed: 

- [x] User Story 10 - User Registration
- [x] User Story 11 - User Registration Missing Details
- [x] User Story 12 - User Registration Must Be Unique

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      session[:user_id] = @user.id
+      flash[:success] = "You are now registered and logged in"
+      redirect_to '/profile'
+    elsif params[:password] != params[:confirm_password]
+      flash[:alert] = "Passwords do not match"
+      render :new
+    else
+      flash.now[:alert] = @user.errors.full_messages.first
+      render :new
+    end
```
Comment (hashmaster3k) [inline app/controllers/users_controller.rb:18]: Looks good!

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  has_secure_password
+  validates_presence_of :name,
+                        :address,
+                        :city,
+                        :state,
+                        :zip
+  validates_presence_of :password, require: true
+  validates :email, uniqueness: true, presence: true
+end
```
Comment (hashmaster3k) [inline app/models/user.rb:10]: Great validation setup.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  def new
+  end
+
+  def create
+    @user = User.new(user_params)
+    if params[:password] == params[:confirm_password] && @user.save
+      session[:user_id] = @user.id
+      flash[:success] = "You are now registered and logged in"
+      redirect_to '/profile'
+    elsif params[:password] != params[:confirm_password]
```
Comment (hashmaster3k) [inline app/controllers/users_controller.rb:12]: Didn't see a test for this. I will make an issue to create a test for this.

**Comment by hashmaster3k** [comment]
Overall looks great!

**Review by hashmaster3k — APPROVED** [review]

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (dcoleman21) [reply app/controllers/users_controller.rb:12]: Do we want to create a column to to-do's?

**Pull request merged**